### ISSUE TITLE
Step-wise simulation

### ISSUE THREAD

**Issue by mortvif** [USEPA/WNTR#204]
Hi! 

I am trying to run a Epanet simulation step by step using the same approach proposed in issue #59. However, step time execution increases linearlly with time in my tests, sugesting that simulation is completly recomputed in each step.  

I haven't found any information on this step-wise approach in the documentation. After a quick code lookup I have found  ENnextH and ENrunH methods in the internal epanet bindings (wntr.epanet.toolkit module) which might be useful. However, it seems to me that they are not used in the exposed high level interface (wntr.sim.epanet module).

Is it possible to perform this step-wise simulation? If so, which is the proper approach to achieve it?

Thanks!

**Comment by kaklise** [comment]
Right now, the step-wise simulation only works with the WNTRSimulator.  We are currently working on features that will extend this to the EpanetSimulator.  This includes the ability to initialize the water network model with simulation results (such as results from a previous timestep).  We plan to push these updates to the dev branch in the next few weeks.

**Comment by mortvif** [comment]
Thanks! I'll stay tuned for the upcoming changes.

**Comment by mortvif** [comment]
Thank you for your quick response!

After analyzing the code I understand the flag that is required is the ``until`` parameter in the ``run_sim`` function. I have adapted your test scenario to run it stepwise as follows (using same setup): 

```python
sim = rt.EpanetSimulator_RT(wn)
prov = rt.RealtimeProvider(timelimit=wn.options.time.duration,
    outfile="temp.out",
    infile="temp.in",
    controls=controls)
for sensor in nodesens:
    sim.add_sensor_instrument(sensor[0], 'node', sensor[1], sensor[2])
for sensor in linksens:
    sim.add_sensor_instrument(sensor[0], 'link', sensor[1], sensor[2])
for controller in controllers:
    sim.add_controller_instrument(controller[0], controller[1], controller[2])

sim.initialize(transmit=prov.proc_sensors, receive=prov.proc_controllers, stop=prov.check_time, file_prefix="temp")
for i in range(1, 101):
    try:
        sim.run_sim(until= i * 3600)      
        print("Simulating until " +str(i*3600) + " in "+str((end-start).seconds) + " seconds")
    except:
        print("Error")

res = sim.close()
```
However, this approach prompts what I understand is an EPANET convergence warning after the first timestep. 
```
/home/mortvif/.local/share/virtualenvs/water_rt-0SEeEPjZ/lib/python3.9/site-packages/wntr/epanet/io.py:2777: UserWarning: Simulation did not converge at time 01:02:00.
warnings.warn('Simulation did not converge at time ' + self._get_time(t) + '.')
```
Execution seems to continue without raising an error but no further timesteps are simulated and results obtained after invoking ``sim.close()`` are limited to those times before the convergence warning.

I don't know if I am missing something or this still requires further work on the library to make it functional. 

Thank you once again for your helpful attention!

**Comment by Naycat** [comment]
Is the step by step simulation for the Epanet simulator now available?

**Comment by kaklise** [comment]
We continue to develop the stepwise and realtime EPANET simulators in WNTR, but they are not quite ready to push to the USEPA main branch.  It will take some time to get these updates fully documented and tested. We can create a separate branch with this capability as soon as it's ready, perhaps this summer.

**Comment by Pu-nk** [comment]
So maybe the step-wise simulator is avaliable now?

**Comment by kaklise** [comment]
The step-wise simulator is not available yet, we are finishing up documentation and tests.

**Comment by rastafaninplakeibol** [comment]
I know this issue is quite dead, but is there any developement about a real-time simulation? @dbhart I can't find the realtime branch you were talking about, neither i can see any similar branch in this repo.

I'm trying to currently develop a real time version of WNRTSimualtor, with ways to add leak, open and close valve etc etc in real time instead of precompute all the controls. If you have anything similar it would be nice to start from a more compatible version

**Comment by Daveonwave** [comment]
@rastafaninplakeibol I don't know if this can help you, but I implemented a step-wise simulation of EPANET in python with another EPANET wrapper, called [Epynet](https://github.com/Vitens/epynet). My repository is [here](https://github.com/Daveonwave/epynet_CPS/tree/main). It is a bit messy because I used it for my master thesis work, but maybe you can take inspiration from that. Right now I'm actually working on a new repository built on this one, cleaning the code and making it more user-friendly.

This may help also @kaklise in implementing the new feature for WNTR too.

**Comment by kaklise** [comment]
@Daveonwave Thanks for sharing the link to your repository, we'll check it out!  Epynet is a great tool! We would really like to finish the stepwise capabilities in WNTR, but it has taken longer than expected. This is certainly high on our list based on user interest.  Let us know if you're interested in contributing and keep us posted on your progress!

**Comment by kaklise** [comment]
@rastafaninplakeibol Thanks for the link to Dyn-WNTR! Contributing to a WNTR stepwise simulator would be great.  Do you plan to migrate this work onto a fork of WNTR?  

@kbonney Can you post a link to the WNTR branch that includes stepwise simulation?

**Comment by kbonney** [comment]
I have some stepwise code on the `stepwise` branch of my WNTR fork: https://github.com/kbonney/WNTR. Feel free to check it out to see if it helps with your needs @mort.

**Comment by rastafaninplakeibol** [comment]
@kaklise yes the original plan was to work on a fork then something went wrong so i just created a new repo to work on it in the meanwhile. I just made the public repo a few days ago to show the code for a paper but i plan to move it to a fork of WNTR as it will be much easier to integrate. 

@kbonney thank you for the reference, i'll dive into it!

**Comment by bertawp31** [comment]
Hi, is the step by step simulation for the Epanet simulator available now?

**Comment by kaklise** [comment]
@bertawp31 The stepwise EpanetSimulator is not yet available on the main branch of WNTR.  I suggest checking out the links to other branches and repositories listed above.  We'll try to coordinate this work into a WNTR release in the future.
